### PR TITLE
Implement Image Coherency

### PR DESCRIPTION
**Pull request by jxy-s**
# Image Coherency
This PR implements the Image Coherency described in https://github.com/processhacker/processhacker/issues/744. Please see that issue for a detailed rundown of the changes.

## Summary
In response to research on process tampering (Hollowing, Doppelganing, Herpaderping). This is an enhancement to Process Hacker which inspects the coherency between an executing process and the image on disk.

For context on Herpaderping please see: https://jxy-s.github.io/herpaderping/

This enhancement adds an API to `phlib` which inspects the image on disk and the image mapped into the process. This calculation represents the similarity between the image on disk and the mapped image. This is displayed in the process tree by "noting" the process' which have "Low Image Coherency". In addition, enhances the process table with a column displaying the "Image Coherency".

### CODE REVIEW

File: phlib/imgcoherency.c [hunk 1]
```diff
@@ -0,0 +1,833 @@
...
+    PPH_PROCESS_IMAGE_COHERENCY_CONTEXT context;
+    PROCESS_BASIC_INFORMATION basicInfo;
+
+    //
+    // This is best-effort context creation, we don't fail if the mapping
+    // fails - the caller of the API may make decisions based on success of
+    // each
+    //
+
+    context = PhAllocateZero(sizeof(PH_PROCESS_IMAGE_COHERENCY_CONTEXT));
```
Comment (dmex) [inline phlib/imgcoherency.c:86]: PhAllocate throws on allocation failure while PhAllocateSafe returns NULL

File: phlib/imgcoherency.c [hunk 2]
```diff
@@ -0,0 +1,833 @@
...
+    // The RVA helper function uses a buffer we supply. For each RVA we will
+    // try to inspect a page - make one
+    //
+    // We only inspect 1 page for each for native code block for simplicity.
+    // Once the native PE is mapped parts the code are "fixed up" by the
+    // loader. It would take substantial work to "unmap" the remote image
+    // to something more coherent and would be time-complex. This gets us
+    // most of the way at the cost accuracy - but it's good enough.
+    //
+    buffer = PhAllocateZero(PAGE_SIZE);
```
Comment (dmex) [inline phlib/imgcoherency.c:291]: Should this be `BYTE buffer[PAGE_SIZE]` instead of allocate?

**Review by dmex — COMMENTED** [review]
LTGM

**Review by dmex — APPROVED** [review]

File: phlib/imgcoherency.c (hunk 2 above)
Comment (dmex) [reply phlib/imgcoherency.c:291]: PhAllocatePage might be more useful

**Comment by valk0** [comment]
Great job!
How does that work with dlls?

**Comment by dmex** [comment]
> How does that work with dlls?

They're identical and use the same PE layout.

**Comment by valk0** [comment]
> > How does that work with dlls?
> 
> They're identical and use the same PE layout.

I meant like which dlls do you compare? Do you recursively check dlls that should be loaded based on the import table on the disk? How does that work with dlls loaded in runtime (LoadLibrary)

**Comment by jxy-s** [comment]
> I meant like which dlls do you compare?

DLLs are not inspected in this implementation. However as @dmex stated, the same calculation may be applied to them. So it may be extended to do so.

**Pull request merged**